src/transaction/log_replication_atomic.cpp

```diff
@@ -413,4 +417,26 @@ namespace cublog
 
     m_locked_classes.emplace (trid, *classoid);
   }
+
+  void
+  atomic_replicator::discard_caches_for_ddl (cubthread::entry &thread_entry, const TRANID trid)
+  {
+    assert (m_locked_classes.count (trid) > 0);
+
+    /* TODO:
+     * This is to update locator_Mht_classnames which contains class names,
+     * and it will be replaced with a function to update specific class name only if needed.
+     */
+    locator_initialize (&thread_entry);
+
+    auto [begin, end] = m_locked_classes.equal_range (trid);
+    for (auto it = begin; it != end; it++)
+      {
+	const OID classoid = it->second;
+
+	heap_classrepr_decache (&thread_entry, &classoid);
+	xcache_remove_by_oid (&thread_entry, &classoid);
+	partition_decache_class (&thread_entry, &classoid);
```

Comment:
Not sure what is the contention on these critical sections on PTS, but - does it make sense to create dedicated functions that take a list of OIDs and decache all of them?

For `heap_classrepr_decache` - critical section `CSECT_HEAP_CHNGUESS` is entered/exited
In the case of `xcache_remove_by_oid`, it's a bit more complicated - a completely new implementation for `xcache_invalidate_entries` is needed that takes a list of OIDs, collects the XASL id's for those that have to deal with the respective OIDs; and then removes the XASL caches
For `partition_decache_class` - critical section `CSECT_PARTITION_CACHE` is entered/exited

For the first and the last, it is relatively easy. For the second one it looks like more work.

Keep in mind that this is done "inline" by the PTS single threaded replication - and we do not know whether we'll be able to ever parallelize this replication.

I am thinking that, for example this de-cache-ing of class OIDs, could be "thrown out" on an async task. Does this make sense? Would it violate consistency on PTS (think - worst case scenario)?

The `locator_initialize` fucntion is even more costly - by retrieving pages from the PS to re-fill the class cache.

Reply:
I agree. I will add optimization issue for this. 
Since it looks like there will be quite a few changes, I'll create a new issue and handle it in parallel!